Use boolean methods for allowed realm types in license state

rjernst opened the pull request:
In xpack the license state contains methods to determine whether a
particular feature is allowed to be used. The one exception is
allowsRealmTypes() which returns an enum of the types of realms allowed.
This change converts the enum values to boolean methods. There are 2
notable changes: NONE is removed as we always fall back to basic license
behavior, and NATIVE is not needed because it would always return true
since we should always have a basic license.

rjernst committed:
Use boolean methods for allowed realm types in license state

In xpack the license state contains methods to determine whether a
particular feature is allowed to be used. The one exception is
allowsRealmTypes() which returns an enum of the types of realms allowed.
This change converts the enum values to boolean methods. There are 2
notable changes: NONE is removed as we always fall back to basic license
behavior, and NATIVE is not needed because it would always return true
since we should always have a basic license.

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authentication)

rjernst committed:
Merge branch 'master' into refactor_license4

rjernst committed:
fix reversed logic

rjernst committed:
Merge branch 'master' into refactor_license4

rjernst committed:
revert saml check

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java:
```diff
@@ -153,17 +137,13 @@ public Realms(Settings settings, Environment env, Map<String, Realm.Factory> fac
         if (licenseState.isAuthAllowed() == false) {
             return Collections.emptyList();
         }
```
While this may never happen in practice, we need to have a consistent view of the license state for correctness here.

```suggestion
        final XPackLicenseState licenseStateSnapshot = copyCurrentLicenseState();
        if (licenseStateSnapshot.isAuthAllowed() == false) {
            return Collections.emptyList();
        }
```

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java:
```diff
@@ -153,17 +137,13 @@ public Realms(Settings settings, Environment env, Map<String, Realm.Factory> fac
+            return standardRealmsOnly;
+        } else {
+            // native realms are basic licensed, and always allowed, even for an expired license
+            return nativeRealmsOnly;
```
Continuation of the previous comment.

```suggestion
        if (licenseStateSnapshot.areAllRealmsAllowed()) {
            return realms;
        } else if (licenseStateSnapshot.areStandardRealmsAllowed()) {
            return standardRealmsOnly;
        } else {
            // native realms are basic licensed, and always allowed, even for an expired license
            return nativeRealmsOnly;
```

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java:
```diff
@@ -128,9 +113,8 @@ public Realms(Settings settings, Environment env, Map<String, Realm.Factory> fac
-        AllowedRealmType allowedRealmType = licenseState.allowedRealmType();
         // If all realms are allowed, then nothing is unlicensed
-        if (allowedRealmType == AllowedRealmType.ALL) {
+        if (licenseState.areAllRealmsAllowed()) {
```
For correctness, this method should also get a snapshot of the license state and possibly even pass it into an internal version of `asList()` that accepts the snapshot.

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java:
```diff
@@ -250,26 +230,24 @@ public void usageStats(ActionListener<Map<String, Object>> listener) {
                             return MapBuilder.<String, Object>newMapBuilder()
                                 .put("enabled", false)
-                                .put("available", isRealmTypeAvailable(allowedRealmType, type))
+                                .put("available", isRealmTypeAvailable(licenseState, type))
```
We'll also want to have a snapshot of the license state here since it can change.

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/oidc/OpenIdConnectBaseRestHandler.java:
```diff
@@ -33,7 +32,7 @@ protected Exception checkFeatureAvailable(RestRequest request) {
         if (failedFeature != null) {
             return failedFeature;
-        } else if (Realms.isRealmTypeAvailable(licenseState.allowedRealmType(), OIDC_REALM_TYPE)) {
+        } else if (licenseState.areStandardRealmsAllowed()) {
```
I do not believe OIDC should be allowed in standard realms based on the `InternalRealms` class. It might be best to have these use Realms.isRealmTypeAvailable to try to keep the logic consolidated?

rjernst replied:
This was an accidental change, I will revert.

jaymode reviewed (changes requested):
I left some comments. I know some of this code was already broken in terms of a license change during method execution, but I think we can easily fix it here.

rjernst committed:
Merge branch 'master' into refactor_license4

rjernst committed:
address feedback

rjernst commented:
@jaymode I believe I have addressed all your feedback. You are right that there are technically race conditions all over the place for our license checking, but I think this is a large enough problem we need to handle it a different way than locally stashing a copy of the license in each part of the codebase. However, I have applied the requested changes here since they were simple enough.

rjernst committed:
test fixes

jaymode reviewed (approved):
LGTM

rjernst committed:
Merge branch 'master' into refactor_license4
